internal/gengapic/genrest_test.go

```diff
@@ -0,0 +1,231 @@
+			MessageType: []*descriptor.DescriptorProto{msg},
+		},
+	}
+	req := plugin.CodeGeneratorRequest{
+		Parameter: proto.String("go-gapic-package=path;mypackage,transport=rest"),
+		ProtoFile: fds,
+	}
+	g.init(&req)
+
+	return mthd, nil
+}
+
+func TestPathParams(t *testing.T) {
+	var g generator
+	g.apiName = "Awesome Mollusc API"
+	g.imports = map[pbinfo.ImportSpec]bool{}
+	g.opts = &options{transports: []transport{rest}}
+
+	for _, tst := range []struct {
+		name     string
+		url      string
+		fields   []string
+		expected map[string]*descriptor.FieldDescriptorProto
+	}{
+		{
+			name:     "no_params",
+			url:      "/kingdom",
+			fields:   []string{"name", "mass_kg"},
+			expected: map[string]*descriptor.FieldDescriptorProto{},
+		},
+		{
+			name:   "params_subset_of_fields",
+			url:    "/kingdom/{kingdom}/phylum/{phylum}",
+			fields: []string{"name", "mass_kg", "kingdom", "phylum"},
+			expected: map[string]*descriptor.FieldDescriptorProto{
+				"kingdom": &descriptor.FieldDescriptorProto{
+					Name:   proto.String("kingdom"),
+					Number: proto.Int32(int32(2)),
+					Type:   typep(descriptor.FieldDescriptorProto_TYPE_INT32),
+				},
+				"phylum": &descriptor.FieldDescriptorProto{
+					Name:   proto.String("phylum"),
+					Number: proto.Int32(int32(3)),
+					Type:   typep(descriptor.FieldDescriptorProto_TYPE_INT32),
+				},
+			},
+		},
+		{
+			name:     "disjoin_fields_and_params",
+			url:      "/kingdom/{kingdom}",
+			fields:   []string{"name", "mass_kg"},
+			expected: map[string]*descriptor.FieldDescriptorProto{},
+		},
+		{
+			name:   "params_and_fields_intersect",
+			url:    "/kingdom/{kingdom}/phylum/{phylum}",
+			fields: []string{"name", "mass_kg", "kingdom"},
+			expected: map[string]*descriptor.FieldDescriptorProto{
+				"kingdom": &descriptor.FieldDescriptorProto{
+					Name:   proto.String("kingdom"),
+					Number: proto.Int32(int32(2)),
+					Type:   typep(descriptor.FieldDescriptorProto_TYPE_INT32),
+				},
+			},
+		},
+		{
+			name:   "fields_subset_of_params",
+			url:    "/kingdom/{kingdom}/phylum/{phylum}/class/{class}",
+			fields: []string{"kingdom", "phylum"},
+			expected: map[string]*descriptor.FieldDescriptorProto{
+				"kingdom": &descriptor.FieldDescriptorProto{
+					Name:   proto.String("kingdom"),
+					Number: proto.Int32(int32(0)),
+					Type:   typep(descriptor.FieldDescriptorProto_TYPE_INT32),
+				},
+				"phylum": &descriptor.FieldDescriptorProto{
+					Name:   proto.String("phylum"),
+					Number: proto.Int32(int32(1)),
+					Type:   typep(descriptor.FieldDescriptorProto_TYPE_INT32),
+				},
+			},
+		},
+	} {
+		mthd, err := setupMethod(&g, tst.url, tst.fields)
+		if err != nil {
+			t.Errorf("test %s setup got error: %s", tst.name, err.Error())
+		}
+
+		actual := g.pathParams(mthd)
+		if diff := cmp.Diff(actual, tst.expected, cmp.Comparer(proto.Equal)); diff != "" {
+			t.Errorf("test %s, got(-),want(+):\n%s", tst.name, diff)
+		}
+	}
+}
+
+func TestQueryParams(t *testing.T) {
```

Comment:
Here it would be worth testing params that are included in the body spec, to make sure they don't show up in the query string.